lib/tests/test_type_hints.c

```diff
@@ -20,180 +20,255 @@
+
 #include "type-hinting.h"
 #include "apphook.h"
+#include <glib.h>
 
+#include <criterion/criterion.h>
+#include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
 #include <math.h>
 
-#define assert_type_hint(hint,expected)                                 \
-  do                                                                    \
-    {                                                                   \
-      TypeHint t;                                                       \
-      GError *e = NULL;                                                 \
-                                                                        \
-      assert_true(type_hint_parse(hint, &t, &e),                        \
-                  "Parsing '%s' as type hint", hint);                   \
-                                                                        \
-      assert_gint(t, expected,                                          \
-                  "Parsing '%s' as type hint results in correct type", hint); \
-    } while(0)                                                          \
-
-static void
-assert_error(GError *error, gint code, const gchar *expected_message)
+typedef struct _StringHintPair
 {
-  assert_not_null(error, "GError expected to be non-NULL");
+  gchar *string;
+  TypeHint value;
+} StringHintPair;
 
-  assert_gint(error->code, code, "GError error code is as expected");
-  if (expected_message)
-    assert_string(error->message, expected_message, "GError error message is as expected");
-}
+typedef struct _StringBoolPair
+{
+  gchar *string;
+  gboolean value;
+} StringBoolPair;
+
+typedef struct _StringDoublePair
+{
+  gchar *string;
+  gdouble value;
+} StringDoublePair;
+
+typedef struct _StringUInt64Pair
+{
+  gchar *string;
+  guint64 value;
+} StringUInt64Pair;
 
-static void
-test_type_hint_parse(void)
+Test(type_hints,test_type_hint_parse)
 {
-  testcase_begin("Testing type hint parsing");
-
-  assert_type_hint(NULL, TYPE_HINT_STRING);
-  assert_type_hint("string", TYPE_HINT_STRING);
-  assert_type_hint("literal", TYPE_HINT_LITERAL);
-  assert_type_hint("boolean", TYPE_HINT_BOOLEAN);
-  assert_type_hint("int", TYPE_HINT_INT32);
-  assert_type_hint("int32", TYPE_HINT_INT32);
-  assert_type_hint("int64", TYPE_HINT_INT64);
-  assert_type_hint("double", TYPE_HINT_DOUBLE);
-  assert_type_hint("datetime", TYPE_HINT_DATETIME);
-  assert_type_hint("default", TYPE_HINT_DEFAULT);
+  TypeHint type_hint;
+  GError *error = NULL;
+
+  const StringHintPair string_value_pairs[] =
+  {
+    {"string",    TYPE_HINT_STRING},
+    {"literal",   TYPE_HINT_LITERAL},
+    {"boolean",   TYPE_HINT_BOOLEAN},
+    {"int",       TYPE_HINT_INT32},
+    {"int32",     TYPE_HINT_INT32},
+    {"int64",     TYPE_HINT_INT64},
+    {"double",    TYPE_HINT_DOUBLE},
+    {"datetime",  TYPE_HINT_DATETIME},
+    {"default",   TYPE_HINT_DEFAULT},
+    {NULL,0} // only for terminating the array
+  };
+
+  int i=0;
+  while (string_value_pairs[i].string)
```

Comment:
There is an option to create Parameterized test case in criterion, it is very similar to what you created but utilize the criterion functions (handles the loop itself).